onadata/apps/api/viewsets/open_data_viewset.py

```diff
@@ -13,25 +15,94 @@
-from onadata.libs.utils.csv_builder import CSVDataFrameBuilder
+from onadata.libs.utils.common_tags import (
+    ATTACHMENTS,
+    NOTES,
+    GEOLOCATION,
+    NA_REP)
 
 BaseViewset = get_baseviewset_class()
 IGNORED_FIELD_TYPES = ['select one', 'select multiple']
 
+# index tags
+DEFAULT_OPEN_TAG = '['
+DEFAULT_CLOSE_TAG = ']'
+DEFAULT_INDEX_TAGS = (DEFAULT_OPEN_TAG, DEFAULT_CLOSE_TAG)
+DEFAULT_NA_REP = getattr(settings, 'NA_REP', NA_REP)
+
 
 def replace_special_characters_with_underscores(data):
     return [re.sub(r"\W", r"_", a) for a in data]
 
 
+def process_tableau_data(data, xform):
+    """
+    Streamlines the row header fields
+    with the column header fields for the same form.
+    Handles Flattenning repeat data for tableau
+    """
+    def get_ordered_repeat_value(key, item, index):
+        """
+        Return Ordered Dict of repeats in the order in which they appear in
+        the XForm.
+        """
+        index_tags = DEFAULT_INDEX_TAGS
+        children = xform.get_child_elements(key, split_select_multiples=False)
+        item_list = OrderedDict()
+        data = {}
+
+        for elem in children:
+            if not question_types_to_exclude(elem.type):
+                new_xpath = elem.get_abbreviated_xpath()
+                item_list[new_xpath] = item.get(new_xpath, DEFAULT_NA_REP)
+                # Loop through repeat data and flatten it
+                # given the key "children/details" and nested_key/
+                # abbreviated xpath "children/details/immunization/polio_1",
+                # generate ["children", index, "immunization/polio_1"]
+                for (nested_key, nested_val) in item_list.items():
+                    xpaths = [
+                        '{key}{open_tag}{index}{close_tag}'.format(
+                            key=nested_key.split('/')[0],
+                            open_tag=index_tags[0],
+                            index=index,
+                            close_tag=index_tags[1])] + [
+                                nested_key.split('/')[1]]
+                    xpaths = "/".join(xpaths)
+                    data[xpaths] = nested_val
+        return data
+
+    result = []
+    if data:
+        headers = xform.get_headers()
+        tableau_headers = remove_metadata_fields(headers)
+        for row in data:
+            diff = set(tableau_headers).difference(set(row))
+            flat_dict = {
+                field: None for field in diff}
+            for (key, value) in row.items():
+                if isinstance(value, list) and key not in [
+                        ATTACHMENTS, NOTES, GEOLOCATION]:
+                    for index, item in enumerate(value, start=1):
+                        # order repeat according to xform order
+                        item = get_ordered_repeat_value(key, item, index)
+                        flat_dict.update(item)
+                else:
+                    flat_dict.update({key: value})
```

Comment:
this should use assignment, since it's a single key and value

Reply:
Sorry, changed that. Thanks